Watchman.Discord/Areas/Protection/Services/MuteService.cs

```diff
@@ -49,42 +50,55 @@ public async Task MuteUserOrOverwrite(Contexts contexts, MuteEvent muteEvent, Us
+            await this._directMessagesService.TrySendMessage(userToMute.Id, x => x.YouHaveBeenMuted(userToMute, muteEvent.TimeRange.End, muteEvent.Reason), contexts);
         }
 
         public async Task<bool> UnmuteIfNeeded(DiscordServerContext server, UserContext userToUnmute)
         {
-            var userMuteEvents = this.GetMuteEvents(server, userToUnmute.Id);
-            var eventToUnmute = this.GetNotUnmutedEvent(userMuteEvents);
+            var userMuteEvents = this.GetUserMuteEvents(server, userToUnmute.Id);
+            var eventToUnmute = userMuteEvents.FirstOrDefault(x => x.Unmuted == false);
             if (eventToUnmute == null)
             {
                 return false;
             }
-            userToUnmute = this._usersService.GetUserById(server, userToUnmute.Id);
-            if (userToUnmute == null) // user could left the server
-            {
-                return false;
-            }
-            await this.UnmuteUser(userToUnmute, eventToUnmute, server);
-            return true;
+            return await this.UnmuteIfNeededSpecificEvent(server, userToUnmute, eventToUnmute);
         }
 
-        public async void UnmuteInFuture(Contexts contexts, MuteEvent muteEvent, UserContext userToUnmute)
+        public async void UnmuteInFuture(Contexts contexts, MuteEvent muteEvent, UserContext userToUnmute, bool sendMessageAfterUnmute = true)
         {
-            var timeRange = muteEvent.TimeRange;
-            await Task.Delay(timeRange.End - timeRange.Start);
-
-            var wasNeededToUnmute = await this.UnmuteIfNeeded(contexts.Server, userToUnmute);
-            if (wasNeededToUnmute)
+            if (muteEvent.TimeRange.End > DateTime.UtcNow)
+            {
+                await Task.Delay(muteEvent.TimeRange.End - DateTime.UtcNow);
+            }
+            var wasNeededToUnmute = await this.UnmuteIfNeededSpecificEvent(contexts.Server, userToUnmute, muteEvent);
+            if (wasNeededToUnmute && sendMessageAfterUnmute)
             {
                 var messagesService = this._messagesServiceFactory.Create(contexts);
                 await messagesService.SendResponse(x => x.UnmutedUser(userToUnmute));
                 await this._directMessagesService.TrySendMessage(userToUnmute.Id, x => x.UnmutedUserForUser(userToUnmute, contexts.Server), contexts);
             }
         }
 
+        public async Task<bool> UnmuteIfNeededSpecificEvent(DiscordServerContext server, UserContext userToUnmute, MuteEvent muteEvent)
+        {
+            var userMuteEvents = this.GetUserMuteEvents(server, userToUnmute.Id);
+            var eventToUnmute = userMuteEvents.FirstOrDefault(x => x.Id == muteEvent.Id);
+            if (eventToUnmute == null || eventToUnmute.Unmuted)
+            {
+                return false;
+            }
+            userToUnmute = this._usersService.GetUserById(server, userToUnmute.Id);
```

Comment:
czemu to pobieramy skoro dostaliśmy na wejściu?
dla gwarancji że użytkownik istnieje?
może tam powinno być coś typu "IsUserExists"

Reply:
możesz dodać IsUserExists do usersService 
może nawet discord ma coś takiego wbudowane i byłoby lepsze wydajnościowo